rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,568 @@
+	OnOperationPlan(op *core.Operation, err error) error
+	// OnPostOperationPlan is called after an operation is planned
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnPostOperationPlan(op *core.Operation, err error) error
+}
+```
+
+## Example Use Cases
+
+- **Advanced GraphQL Operation Handling**: A module that walks the parsed operation and performs custom logic based on the operation type, fields, and arguments.
+- **Request validation**: A module that validates incoming requests and returns an error if the request is invalid.
+- **Custom Telemetry**: A module that creates custom spans or metric data and sends it to a telemetry backend.
+- **Custom Authentication / Authorization**: A module that authenticates incoming requests and adds user information to the subgraph requests.
+- **Response interception**: A module that intercepts responses from subgraphs and modifies them before they are sent to the client.
+- **Response Caching**: A module that caches responses from subgraphs and returns them for identical requests.
+- **GraphQL Directive Handling**: A module that reacts to specific GraphQL directives in the operation definition.
+- **Enriching Logs**: A module that adds custom log fields to the gateway and subgraph logs.
+
+## Backwards Compatibility
+
+The new module system is not backwards compatible with the old module system. Existing custom modules will need to be rewritten to use the new interfaces and API. We will provide a migration guide and tooling to help developers migrate their custom modules to the new system.
+
+# Example Modules
+
+__All examples are pseudocode and not tested but they are as close as possible to the final implementation__
+
+## Custom Telemetry
+
+This module adds custom attributes to the OpenTelemetry span for each gateway request. Data can come from the request, the gateway configuration, or external sources. The example modifies the first span of the gateway but depending on the hook, the span is different.
+
+```go
+type MyModule struct{}
+
+// Ensure that MyModule implements the GatewayHooks interface
+var _ GatewayHooks = (*MyModule)(nil)
+
+func (m *MyModule) OnGatewayRequest(req *core.GatewayRequest, err error) error {
+	req.Telemetry.Span.AddEvent("Gateway Request")
+	req.Telemetry.Span.AddAttributes(
+		attribute.String("gateway.name", "myGateway"),
+		attribute.String("gateway.version", "v1"),
+	)
+	return nil
+}
+```
+
+## Intercepting Gateway Responses
+
+This module intercepts the final Gateway response to rewrite errors and add custom extensions. You have access to the parsed operation, the gateway request, and the gateway response.
+
+```go
+type MyModule struct{}
+
+// Ensure that MyModule implements the GatewayHooks interface
+var _ GatewayHooks = (*MyModule)(nil)
+
+func (m *MyModule) OnGatewayResponse(res *core.GatewayResponse, err error) error {
+	// Add custom extensions to the response
+	res.Response.Extensions["myExtension"] = "myValue"
```

Comment:
We don't currently expose the response as an object that could be used like this, it's just a byte slice.
This is for performance reasons. We're encoding the response into a byte buffer.
However, it's possible that we would expose "data" as a byte slice, "errors" as a fastjson array, and extensions as a fastjson object. This way, you could have some level of access, probably enough for the required customizations.
I assume that most customizations won't touch data, but errors and extensions, which we can easily expose.

However, if we do it like this, I'd restructure the hook here.
OnGatewayResponse will really "just" contain the encoded response.
We can add another hook, e.g. "PreGraphQLResponseEncoding" which gives you all the parts (data, errors, extensions) before we encode them as a single response.

Reply:
I understand your concerns and I like the fastjson object idea. I think we should try to make the API as intuitive as possible. In the first draft, I'd like to avoid dedicated hooks. We should reiterate on this before the release once we did some performance tests. We could also provide optimized helper functions to get and manipulate extensions and errors that operating on byte level.